File: python/adbconnect.py
# Fichier gérant la connexion entre l'appareil executant le script et l'appareil android.

import subprocess
import re

device = ""

def adb_cmd(cmd):
    if not device:
        print("[adbconnect/simulation] La commande ne peut pas être exécutée")
        return

    combined_cmd = ["adb", "-s", device] + cmd.split()

    return subprocess.check_output(combined_cmd)

def choose_device():
    global device
    list_devices = []

    out = subprocess.check_output(["adb", "devices"], text=True)
    pattern = r"([^\r\n\t]+)\t([^\s]+)"

    print("[adbconnect] List of devices attached")

    matches = re.findall(pattern, out)

    if matches:
        for i, m in enumerate(matches):

            if m[1] == "device":
                print(f"[adbconnect] [{len(list_devices)}] ", end="")
                list_devices.append(m[0])
            else:
                print("[adbconnect] [X] ", end="")
            
            print(f"{m[0]}\t{m[1]}")

        output = input("\n[adbconnect] Quel appareil voulez vous utiliser [digit or \'none\'] ? ")
        if output == "none":
            print("[adbconnect] Aucun appareil connecté\n[adbconnect] Mode simulation activé")
        else:
            id_device = int(output)
            device = list_devices[id_device]
            print("[adbconnect] Connecté à : " + device)
    else:
        print("[adbconnect] Aucun appareil utilisable atuellement...\n")
        output = input("[adbconnect] Voulez vous continuer en mode simulation [Y/n] ? ")
        if output == "Y" or output == "y":
            print("[adbconnect] Aucun appareil connecté\n[adbconnect] Mode simulation activé")
        else: exit()

def capture_screen(parentFolder):
    if not device:
        name = input("[adbconnect/simulation] Quel fichier local souhaitez vous ouvrir ? ")
        
        if name == "" or name == "\n" or name == "\r\n":
            name = "screen.png"

        return name

    else:
        name = input("[adbconnect] Sous quel nom souhaitez vous enregistrer la capture ? ")

        if name == "" or name == "\n" or name == "\r\n":
            name = "screen.png"

        data = adb_cmd("exec-out screencap -p")

        if data:
            with open(parentFolder + name, "wb") as f:
                f.write(data)

        print("[adbconnect] Image enregistrée dans : " + parentFolder + name)

    return name

def click(x, y):
    if not device:
        print(f"[adbconnect/simulation] clic x:{x} y:{y}")
        return

    else:
        cmd = "shell input tap " + str(x) + " " + str(y)
        adb_cmd(cmd)
        print(f"[adbconnect] clic x:{x} y:{y}")
        return

File: cpp/adbconnect.cpp
#include "adbconnect.h"

#include <iostream>
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <regex>
#include <vector>

using namespace std;
namespace fs = std::filesystem;

string ADBConnect::command(const string& cmd) {
    system((cmd + " > output.txt").c_str());
    
    ifstream file("output.txt");
    string output((istreambuf_iterator<char>(file)),
                       istreambuf_iterator<char>());

    remove("output.txt");
    return output;
}

string ADBConnect::adb_command(const string& device, const string& cmd) {
    if (device == "") {
        cout << "[adbconnect/simulation] La commande ne peut pas être exécutée" << endl;
        return "";
    }

    const string full_command = "adb -s " + device + " " + cmd;
    return command(full_command);
}

string ADBConnect::choose_device() {
    string device = "";
    vector<string> list_devices;

    const string out = command("adb devices");
    const regex pattern(R"(([^\r\n\t]+)\t([^\s]+))");

    cout << "[adbconnect] List of devices attached" << endl;

    sregex_iterator next(out.begin(), out.end(), pattern);
    sregex_iterator end;

    while (next != end) {
        smatch match = *next;
        if (match[2].str() == "device") {
            cout << "[adbconnect] [" << list_devices.size() << "] ";
            list_devices.push_back(match[1].str());
        }
        else {
            cout << "[adbconnect] [X] ";
        }
        cout << match[1].str()
             << '\t'
             << match[2].str()
             << endl;
            
        next++;
    }

    if ((int)list_devices.size() > 0) {
        string response = "";
        cout << endl << "[adbconnect] Quel appareil voulez vous utiliser [digit or \'none\'] ? ";
        cin >> response;
        if (response == "none") {
            cout << "[adbconnect] Aucun appareil connecté" << endl << "[adbconnect] Mode simulation activé" << endl;
        }
        else {
            const int id_device = stoi(response);
            device = list_devices.at(id_device);
            
            cout << "[adbconnect] Connecté à : " << device << endl;
        }
    }
    else {
        cout << "[adbconnect] Aucun appareil utilisable atuellement..."
             << endl << endl
             << "[adbconnect] Voulez vous continuer en mode simulation [Y/n] ? ";
        string response = "";
        cin >> response;
        if (response == "y" or response == "Y") {
            cout << "[adbconnect] Aucun appareil connecté" << endl << "[adbconnect] Mode simulation activé" << endl;
        }
        else {
            exit(0);
        }
    }

    return device;
}

string ADBConnect::captureScreen(const string& device, const string& parentfolder) {
    if (device == "") {    
        string name = "";
        cout << "[adbconnect/simulation] Quel fichier local souhaitez vous ouvrir ? ";
        cin >> name;

        if (name == "" or name == "\n" or name == "\r\n") {
            name = "screen.png";
        }

        return name;
    }
    else {
        string name = "";
        cout << "[adbconnect] Sous quel nom souhaitez vous enregistrer la capture ? ";
        cin >> name;

        if (name == "" or name == "\n" or name == "\r\n") {
            name = "screen.png";
        }

        const string cmd = "adb -s " + device + " exec-out screencap -p";
        FILE* pipe = _popen(cmd.c_str(), "rb");
        ofstream f(parentfolder + name, ios::binary);

        char buffer[4096];
        size_t n;

        while ((n = fread(buffer, 1, sizeof(buffer), pipe)) > 0) {
            f.write(buffer, n);
        }
        
        pclose(pipe);
        f.close();

        cout << "[adbconnect] Image enregistrée dans : " << parentfolder + name << endl;
        return name;
    }
}

void ADBConnect::click(const string& device, const int& x, const int& y) {
    if (device == "") {
        cout << "[adbconnect/simulation] clic x:"
             << to_string(x)
             << " y:"
             << to_string(y)
             << endl;
        return;
    }

    const string cmd = "shell input tap " + to_string(x) + " " + to_string(y);
    adb_command(device, cmd);
    cout << "[adbconnect] clic x:"
             << to_string(x)
             << " y:"
             << to_string(y)
             << endl;

    return;
}